app/models/user.rb

```diff
@@ -372,6 +374,53 @@ def only_card_grant_user?
     card_grants.size >= 1 && events.size == 0
   end
 
+  def backup_codes_enabled?
+    active_backup_codes.any?
+  end
+
+  def generate_backup_codes!
+    backup_codes.previewed.destroy_all
+
+    codes = []
+    pepper = Credentials.fetch(:BACKUP_CODE_PEPPER)
+    while codes.size < 10
+      code = SecureRandom.alphanumeric(10)
+      next if codes.include?(code)
+
+      salt = Base64.strict_encode64(SecureRandom.random_bytes(64))
+      begin
+        # backup code pepper must be at least 32 bytes
+        backup_codes.create!(code_hash: User::BackupCode.gen_hash(code:, salt:, pepper:), salt: salt)
+      rescue ActiveRecord::RecordInvalid
+        # if the code is already in use, skip it
+        next
+      end
+      codes << code
+    end
+
+    codes
```

Comment:
1. **concern**: I find it surprising that this logic lives in the `User` model. Might I suggest we add a class method on `User::BackupCode` that generates the code, saves the record, and returns both the code and the record?
2. **concern**: What's our reasoning here for rolling our own crypto? Can we make use of `has_secure_password` (https://api.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html#method-i-has_secure_password) instead?
3. **concern**: This is tricky logic we should definitely have tests for. Happy to help if need be.

Reply:
@davidcornu would you mind setting up a pairing session with @rluodev on these three points?